source/common/http/filter/lua/lua_filter.cc

```diff
@@ -0,0 +1,490 @@
+    ENVOY_LOG(debug, "buffering body");
+    return FilterDataStatus::StopIterationAndBuffer;
+  } else if (state_ == State::Responded) {
+    return FilterDataStatus::StopIterationNoBuffer;
+  } else {
+    headers_continued_ = true;
+    return FilterDataStatus::Continue;
+  }
+}
+
+FilterTrailersStatus StreamHandleWrapper::onTrailers(HeaderMap& trailers) {
+  ASSERT(!end_stream_);
+  end_stream_ = true;
+  trailers_ = &trailers;
+
+  if (state_ == State::WaitForBodyChunk) {
+    ENVOY_LOG(debug, "resuming nil body chunk due to trailers");
+    state_ = State::Running;
+    coroutine_->resume(0, yield_callback_);
+  } else if (state_ == State::WaitForBody) {
+    ENVOY_LOG(debug, "resuming body due to trailers");
+    state_ = State::Running;
+    coroutine_->resume(luaBody(coroutine_->luaState()), yield_callback_);
+  }
+
+  if (state_ == State::WaitForTrailers) {
+    // Mimic a call to trailers which will push the trailers onto the stack and then resume.
+    state_ = State::Running;
+    coroutine_->resume(luaTrailers(coroutine_->luaState()), yield_callback_);
+  }
+
+  FilterTrailersStatus status = (state_ == State::HttpCall || state_ == State::Responded)
+                                    ? FilterTrailersStatus::StopIteration
+                                    : FilterTrailersStatus::Continue;
+
+  if (status == FilterTrailersStatus::Continue) {
+    headers_continued_ = true;
+  }
+
+  return status;
+}
+
+int StreamHandleWrapper::luaRespond(lua_State* state) {
+  ASSERT(state_ == State::Running);
+
+  luaL_checktype(state, 2, LUA_TTABLE);
+  size_t body_size;
+  const char* raw_body = luaL_optlstring(state, 3, nullptr, &body_size);
+  HeaderMapPtr headers = buildHeadersFromTable(state, 2);
+
+  uint64_t status;
+  if (headers->Status() == nullptr ||
+      !StringUtil::atoul(headers->Status()->value().c_str(), status) || status < 200 ||
+      status >= 500) {
+    luaL_error(state, ":status must be between 100-599");
+  }
+
+  Buffer::InstancePtr body;
+  if (raw_body != nullptr) {
+    body.reset(new Buffer::OwnedImpl(raw_body, body_size));
+    headers->insertContentLength().value(body_size);
+  }
+
+  // Once we respond we treat that as the end of the script even if there is more code. Thus we
+  // yield.
+  callbacks_.respond(std::move(headers), body.get(), state);
+  state_ = State::Responded;
+  return lua_yield(state, 0);
+}
+
+HeaderMapPtr StreamHandleWrapper::buildHeadersFromTable(lua_State* state, int table_index) {
+  HeaderMapPtr headers(new HeaderMapImpl());
+
+  // Build a header map to make the request. We iterate through the provided table to do this and
+  // check that we are getting strings.
+  lua_pushnil(state);
+  while (lua_next(state, table_index) != 0) {
+    // Uses 'key' (at index -2) and 'value' (at index -1).
+    const char* key = luaL_checkstring(state, -2);
+    const char* value = luaL_checkstring(state, -1);
+    headers->addCopy(LowerCaseString(key), value);
```

Comment:
Memory resource use seems to be another place that multi-tenant Lua filters will fall flat..